src/dfmpreview/src/extension.ts

```diff
@@ -0,0 +1,273 @@
+    return document.languageId === 'markdown'
+        && document.uri.scheme !== 'markdown'; // prevent processing of own documents
+}
+
+
+function getMarkdownUri(uri: Uri) {
+    return uri.with({ scheme: 'markdown', path: uri.path + '.rendered', query: uri.toString() });
+}
+
+
+
+function showPreview(dfm_preview: PreviewCore, uri?: Uri, sideBySide: boolean = false) {
+    dfm_preview._is_firsttime = true;
+    let resource = uri;
+    if (!(resource instanceof Uri)) {
+        if (window.activeTextEditor) {
+            // we are relaxed and don't check for markdown files
+            resource = window.activeTextEditor.document.uri;
+        }
+    }
+
+    if (!(resource instanceof Uri)) {
+        if (!window.activeTextEditor) {
+            // this is most likely toggling the preview
+            return commands.executeCommand('markdown.showSource');
+        }
+        // nothing found that could be shown or toggled
+        return;
+    }
+
+    let thenable = commands.executeCommand('vscode.previewHtml',
+        getMarkdownUri(resource),
+        getViewColumn(sideBySide),
+        `Preview '${path.basename(resource.fsPath)}'`);
+
+    document_uri = getMarkdownUri(resource);
+    dfm_preview.callDfm();
+    return thenable;
+}
+
+function getViewColumn(sideBySide): ViewColumn {
+    const active = window.activeTextEditor;
+    if (!active) {
+        return ViewColumn.One;
+    }
+
+    if (!sideBySide) {
+        return active.viewColumn;
+    }
+
+    switch (active.viewColumn) {
+        case ViewColumn.One:
+            return ViewColumn.Two;
+        case ViewColumn.Two:
+            return ViewColumn.Three;
+    }
+
+    return active.viewColumn;
+}
+
+function showSource(mdUri: Uri) {
+    if (!mdUri) {
+        return commands.executeCommand('workbench.action.navigateBack');
+    }
+
+    const docUri = Uri.parse(mdUri.query);
+
+    for (let editor of window.visibleTextEditors) {
+        if (editor.document.uri.toString() === docUri.toString()) {
+            return window.showTextDocument(editor.document, editor.viewColumn);
+        }
+    }
+
+    return workspace.openTextDocument(docUri).then(doc => {
+        return window.showTextDocument(doc);
+    });
+}
+
+//this class is to call the dfmserver(child_process) by Interprocess communication
+class PreviewCore {
+    private _spawn: child_process.ChildProcess;
+    private _waiting: boolean;
+    public _is_firsttime: boolean;
+
+    constructor(context: ExtensionContext) {
+        let extpath = context.asAbsolutePath('./DfmParse/PreviewCore.exe');
+        this._spawn = child_process.spawn(extpath);
+        this._waiting = false;
+
+        this._spawn.stdout.on('data', function (data) {
+            let tmp = data.toString();
+            let endcharcode = tmp.charCodeAt(tmp.length - 1);
+            if (Is_end && endcharcode == ENDCODE) {
+                previewresult = tmp;
+                provider.update(document_uri);
```

Comment:
How about passing `provider` into the constructor instead of using global variables?